File: NTP-Miner/Python/ntp_df.py
import time
from tkinter import END


def binary_search2(fre, cand, level):
    low, high = 0, len(fre) - 1
    if low > high:
        return -1
    while low <= high:
        mid = int((low + high) / 2)
        if cand == fre[mid][0:level - 1]:
            slow, shigh = low, mid
            if cand == fre[low][0:level - 1]:
                return low
            else:
                while slow < shigh:  # 此处二分有修改
                    smid = int((slow + shigh) / 2)
                    if cand == fre[smid][0:level - 1]:
                        shigh = smid
                    else:
                        slow = smid + 1
                if slow < 0 or slow >= len(fre):
                    slow = 0
                return slow
        elif cand < fre[mid][0:level - 1]:
            high = mid - 1
        else:
            low = mid + 1
    return -1


def binary_search(fre, cand, level):
    low, high = 0, len(fre) - 1
    if low > high:
        return -1
    while low < high:
        mid = int((low + high) / 2)
        if cand <= fre[mid][0:level - 1]:
            high = mid
        else:
            low = mid + 1
    if cand == fre[low][0:level - 1]:
        return low
    elif low + 1 < len(fre) and cand == fre[low + 1][0:level - 1]:
        return low + 1
    else:
        return -1


def read_file(file_path):
    file = open(file_path, 'r')
    return file.readlines()


# 通过模式连接生成候选模式
def gen_candidate(fre, level, fre0):
    candidate = []
    for i in fre:
        for j in fre0:
            cand = i[0:level] + j
            candidate.append(cand)
    return candidate


class NTP_df:
    output_filepath = ""
    output_filename = "NTP-df-Output.txt"
    mingap, maxgap = 0, 3
    minsup = 50
    s = "hilkmftwv"
    m = "rcqgpsyn"
    w = "adeuox"
    sDB = []
    sub_ptn = []
    ptn_len = 0
    frequent = {}
    compnum = 0

    def __init__(self, file_path, output_filepath, strong, middle, week, min_gap, max_gap, min_sup):
        self.sDB = read_file(file_path)
        self.output_filepath = output_filepath
        self.mingap, self.maxgap = min_gap, max_gap
        self.minsup = min_sup
        self.s = strong
        self.m = middle
        self.w = week
        self.sub_ptn = []
        self.frequent = {}
        self.ptn_len = 0
        self.compnum = 0

    class sub_ptn_struct:
        start = ''
        end = ''
        min, max = 0, 0

        def __init__(self, start, end, min, max):
            self.start = start
            self.end = end
            self.min = min
            self.max = max

    def min_freItem(self):
        counter = {}
        fre = []
        for strs in self.sDB:
            for c in strs:
                if self.belong(c, self.s) or self.belong(c, self.m):
                    if counter.get(c):
                        counter[c] = counter[c] + 1
                    else:
                        counter[c] = 1
        for key in counter.keys():
            if counter[key] >= self.minsup:
                fre.append(key)
        return sorted(fre)

    def deal_range(self, pattern):
        sub_ptn = []
        self.ptn_len = 0
        if len(pattern) == 1:
            sub_ptn.append(self.sub_ptn_struct(pattern[0], "", 0, 0))
        for i in range(0, len(pattern) - 1):
            sub_ptn.append(self.sub_ptn_struct(pattern[i], pattern[i + 1], self.mingap, self.maxgap))
            self.ptn_len += 1
        return sub_ptn

    def create_nettree(self, nettree, seq):
        occurnum = 0
        for i in range(0, self.ptn_len + 1):
            nettree.append([])
        for i in range(0, len(seq) - self.ptn_len):
            if seq[i] != self.sub_ptn[0].start:
                continue
            nettree[0].append(i)
            occurnum = occurnum + self.create_subnettree(nettree, seq, i, 2)
        return occurnum

    def create_subnettree(self, nettree, seq, parent, L):
        if L > self.ptn_len + 1:
            return 1
        for i in range(parent + 1, parent + self.sub_ptn[L - 2].min + 1):
            if self.belong(seq[i], self.s):
                return 0
        for i in range(parent + self.sub_ptn[L - 2].min + 1, parent + self.sub_ptn[L - 2].max + 2):
            # print(i, len(seq), parent)
            if i >= len(seq):
                break
            if seq[i] == self.sub_ptn[L - 2].end:
                k = len(nettree[L - 1])
                flag = -1
                for j in range(k):
                    if i == nettree[L - 1][j]:
                        flag = j
                        break
                if flag == -1:
                    nettree[L - 1].append(i)
                    if self.create_subnettree(nettree, seq, i, L + 1):
                        return 1
            if not self.belong(seq[i], self.m) and not self.belong(seq[i], self.w):
                break
        return 0

    def belong(self, ch, str):
        for c in str:
            if ch == c:
                return True
        return False

    def CalNtp(self, p, strs):
        self.sub_ptn = self.deal_range(p)
        num = 0
        if(self.ptn_len+1 > len(strs)):
            num = 0
        else:
            nettree = []
            num = self.create_nettree(nettree, strs)
        return num

    def mineFre(self, p, num, fre):
        self.frequent[p] = num
        for e in fre:
            q = p + e
            self.compnum += 1
            num = 0
            for t in self.sDB:
                if len(t) > 0:
                    num += self.CalNtp(q, t)
                if num >= self.minsup:
                    break
            if num >= self.minsup:
                self.mineFre(q, num, fre)

    def output(self):
        output_file = open(self.output_filepath + "/" + self.output_filename, 'w')
        for key in self.frequent.keys():
            output_file.write(key + ":" + str(self.frequent[key]) + "\t")

    def solve(self, text):
        text.insert(END, "NTP_df算法开始运行，请等待...\n")
        text.insert(END, "算法参数：\n")
        text.insert(END, "强字符集：" + self.s + " 中字符集：" + self.m + " 弱字符集：" + self.w + "\n")
        text.insert(END,
                    "mingap:" + str(self.mingap) + " maxgap:" + str(self.maxgap) + " minsup:" + str(self.minsup) + "\n")
        f_level = 1
        freArr = []

        begin_time = time.time()
        fre = self.min_freItem()

        for p in fre:
            num = 0
            rest = 0
            for strs in self.sDB:
                rest = self.minsup - num
                if len(strs) > 0:
                    num += self.CalNtp(p, strs)
                if num >= self.minsup:
                    break
            self.mineFre(p, num, fre)
        end_time = time.time()
        time_consuming = end_time - begin_time
        freNum = 0
        text.insert(END, "NTP_df算法运行完毕。\n")
        for key in self.frequent.keys():
            text.insert(END, key + ":" + str(self.frequent[key]) + "\t")
        text.insert(END, "\nThe number of frequent patterns:" + str(len(self.frequent)) + "\n")
        text.insert(END, "The time-consuming:" + str(time_consuming * 1000) + "ms" + "\n")
        text.insert(END, "The number of calculation:" + str(self.compnum) + "\n")
        self.output()
        text.insert(END, "挖掘结果已写入：" + self.output_filepath + "/" + self.output_filename + "\n")

    def solve_test(self):
        print("NTP_df算法开始运行，请等待...")
        print("算法参数：")
        print("强字符集：" + self.s + " 中字符集：" + self.m + " 弱字符集：" + self.w)
        print("mingap:" + str(self.mingap) + " maxgap:" + str(self.maxgap) + " minsup:" + str(self.minsup))
        f_level = 1
        freArr = []

        begin_time = time.time()
        fre = self.min_freItem()

        for p in fre:
            num = 0
            rest = 0
            for strs in self.sDB:
                rest = self.minsup - num
                if len(strs) > 0:
                    num += self.CalNtp(p, strs)
                if num >= self.minsup:
                    break
            self.mineFre(p, num, fre)
        end_time = time.time()
        time_consuming = end_time - begin_time
        freNum = 0
        print("NTP_df算法运行完毕。")
        for key in self.frequent.keys():
            print(key + ":" + str(self.frequent[key]) + "\t", end="")
        print("\nThe number of frequent patterns:" + str(freNum))
        print("The time-consuming:" + str(time_consuming * 1000) + "ms")
        print("The number of calculation:" + str(self.compnum))


if __name__ == '__main__':
    print("请输入输入源文件名（包含路径）：")
    file = input()
    print("请输入强字符集：")
    s = input()
    print("请输入中字符集：")
    m = input()
    print("请输入弱字符集：")
    w = input()
    print("请输入min_gap：")
    min_gap = input()
    print("请输入max_gap：")
    max_gap = input()
    print("请输入min_sup：")
    min_sup = input()
    NTP_df(file, s, m, w, min_gap, max_gap, min_sup).solve_test()
    # NTP_Miner("D:\\github\\Pattern-Mining\\NTP-Miner\\NTP-DataSet\\SDB6.txt", "hilkmftwv", "rcqgpsyn", "adeuox", 0, 3, 400).solve_test()

File: NTP-Miner/C++/NTP-df.cpp
#include <map>
#include <set>
#include <vector>
#include <string>
#include <iostream>
#include <stdlib.h>
#include <windows.h>
#include <fstream>
#include <stack>
using namespace std;

#define K 6000   //The sequence number of sequence database
#define M 1000   //The length of pattern
#define N 20000  //The length of sequence
char S[N];  //sequence
int ptn_len=0;  
int seq_len=0;

int mingap,maxgap; //gap constraint

struct sub_ptn_struct   //a[0,3]c => start[min,max]end
{
	char start,end;		
	int min,max;
};

sub_ptn_struct sub_ptn[M];  //pattern p[i]

struct seqdb                 
{
	int id;                  // sequence id
	char S[N];              // sequence
} sDB[K];                 // sequence database
int minsup; 
int nn=0; 
int NumbS;
///////////////////////////
string item[30];
map <string, int> frequent;  //frequent pattern and its support

int compnum = 0, frenum = 0;  //compute number and frequents number
int occurnum=0;
/////


/*

char mw[15]="BbCcoaO";
char sm[19]="DdEeFfBbCc";
char m[20]="BbCc";


char mw[24]="hilkmftwvrcqgpsynadeuox";
char sm[24]="hilkmftwvrcqgpsynadeuox";
char m[24]="hilkmftwvrcqgpsynadeuox";

char mw[20]="rcqgpsynadeuox";
char sm[20]="hilkmftwvrcqgpsyn";
char m[10]="rcqgpsyn";
char s[10]="hilkmftwv";

char mw[20]="AaBbCcDdEeFfGg";
char sm[20]="EeFfGgHhIiJj";
char m[10]="EeFfGg";
char s[10]="HhIiJj";
*/

char mw[24]="hilkmftwvrcqgpsynadeuox";
char sm[24]="hilkmftwvrcqgpsynadeuox";
char m[24]="hilkmftwvrcqgpsynadeuox";
char s[10]="";

/////
void deal_range(string pattern)      
//put sub-pattern "a[1,3]b" into sub_ptn and sub_ptn.start=a，sub_ptn.end=b, sub_ptn.min=1，sub_ptn.max=3
{
	ptn_len=0;
	memset(sub_ptn, 0, sizeof(sub_ptn_struct));
	char p[M];
	strcpy(p,pattern.c_str()); 
	if (strlen(p)==1)
	{
		sub_ptn[ptn_len].start = p[0];
		sub_ptn[ptn_len].max =sub_ptn[ptn_len].min=0;
	}
	for(int i=0;i<strlen(p)-1;i++ )
	{
		sub_ptn[ptn_len].start =p[i];
		sub_ptn[ptn_len].end =p[i+1];
		sub_ptn[ptn_len].min=mingap;	
		sub_ptn[ptn_len].max=maxgap;
		ptn_len++;
	}
}



bool belong(char c,char p[])
{
	
	for(int i=0;i<strlen(p);i++)
	{
		if(c==p[i])
		{
			return true;
		}
	}
	return false;

}




void min_freItem()
{
	map <string, int > counter; 
	string mine;
	for(int t = 0; t < NumbS; t++)  //count item and its support num
	{
		strcpy(S,sDB[t].S);
		for(int i=0;i<strlen(S);i++)
		{
			if ((S[i]>='a'&&S[i]<='z')||(S[i]>='A'&&S[i]<='Z'))
			{
				if(belong(S[i],sm))
				{
					mine=S[i];
					counter[mine]++;
				}
			
			}	
		}
	}
	nn = 0;
	for (map <string,int >::iterator the_iterator = counter.begin (); the_iterator!= counter.end (); )
	{
		if (the_iterator->second < minsup)
		{
			map <string,int >::iterator tmp = the_iterator;
			++tmp;
			counter.erase (the_iterator);
			the_iterator = tmp;
		} 
		else
		 {
			 item[nn++] = the_iterator->first.at(0);
			 ++the_iterator;
		 }
	}
}
int create_subnettree(vector <int> *nettree,int parent,int L)
{
	int j=0;
	if(L>ptn_len+1)
	{
		return 1;
	}
	int flag=0;
	for(j=parent+1;j<=parent+sub_ptn[L-2].min;j++)
	{
		if(belong(S[j],s))
		{
			return 0;
		}
	}
	

	for(j=parent+sub_ptn[L-2].min+1;j<=parent+sub_ptn[L-2].max+1;j++)
	{	
		if(S[j]==sub_ptn[L-2].end)
		{
			int k=nettree[L-1].size();
			flag=-1;
			for(int i=0;i<k;i++)
			{
				if(nettree[L-1][i]==j)
				{
					flag=i;
					break;
				}
			}

			if(flag==-1)
			{
				nettree[L-1].push_back(j);
				
				if(create_subnettree(nettree,j,L+1))
				{
					return 1;
					//break;
				}
		
			}
		}///创建当前节点

		if(!belong(S[j],mw))
		{
			break;
		}
	}
	return 0;

}

void create_nettree(vector <int> *nettree)
{	
	occurnum=0;
	for (int i=0;i<ptn_len+1;i++)
		nettree[i].clear();//网树层初始化
	
	for (i=0;i<seq_len-ptn_len;i++)
	{
		if(S[i]!=sub_ptn[0].start)
		{
			continue;
		}
		nettree[0].push_back(i);
		occurnum+=create_subnettree(nettree,i,2);
		
	}
	
}




int CalNtp(string p)
{
	deal_range(p);
	int num=0;
	if(ptn_len+1 > seq_len)
	{
		num=0;
	}else{
		vector <int> *nettree;
		nettree=new vector <int> [ptn_len+1];
		
		create_nettree(nettree);
		num=occurnum;
		delete []nettree;
	}
	return num;
}


void mineFre(string p,int num)
{
	frequent[p] = num;  //add to frequent array
	for(int e = 0; e < nn; e++)
	{
		string q = p + item[e];
		compnum++;
		num = 0;
		for(int t = 0; t < NumbS; t++)
		{
			if(strlen(sDB[t].S) > 0)
			{
				strcpy(S,sDB[t].S);
				seq_len=strlen(S);
				num += CalNtp(q);
			}
			if (num>=minsup)
				break;
		}
		if (num>=minsup)
			mineFre(q,num);	

	}
}

void read_file()
{
	fstream file;
	char filename[256];
	cout <<"---------------------------------------------------------------------------\nPlease input file name:";
	string buff;
	file.open("F:/NTP-Miner/NTP-DataSet/SDB1.txt",ios::in);   //open sequence file
	int i=0;
	while(getline(file, buff))
	{
		strcpy(sDB[i].S, buff.c_str());
		i++;
	}
	NumbS=i;
	for(int t = 0; t < NumbS; t++)
	{
		sDB[t].id = t + 1;
		//cout << "\t(" << sDB[t].id << ",(" << sDB[t].S << "))" << endl;
	}
}


void main()
{
	read_file();
	cout<<"input the mingap,maxgap:"<<endl;
	cin>>mingap>>maxgap;
	cout<<"input the minsup:"<<endl;
	cin>>minsup;
	DWORD begintime=GetTickCount();
	min_freItem();   //cout frequent items and store item[]
	for(int e = 0; e < nn; e++)    // each item
	{
		string p = item[e];
		int num = 0;
		int rest=0;
		for(int t = 0; t < NumbS; t++)
		{
			rest=minsup-num;

			if(strlen(sDB[t].S) > 0)
			{
				strcpy(S,sDB[t].S);
				num += CalNtp(p);
			}
			if(num >= minsup)
			{
				break;
			}
		}
			mineFre(p,num);    // mining all frequent patterns
	}
	DWORD endtime=GetTickCount();
	cout<<"The number of frequent patterns:"<<frequent.size()<<endl;
	for(map<string,int> ::iterator f_iterator = frequent.begin (); f_iterator!= frequent.end ();f_iterator++)
	{
		cout<< f_iterator->first<<":";
		cout<< f_iterator->second<<"\t";
	}
	cout<<endl;
	cout<<"The number of frequent patterns:"<<frequent.size()<<endl;
	cout <<"The time-consuming:"<<endtime-begintime<<"ms. \n";
	cout <<"The number of calculation:"<<compnum<<" \n";
	
}
